src/execution_plan/ops/op_semi_apply.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2018-2019 Redis Labs Ltd. and Contributors
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#include "op_semi_apply.h"
+#include "../execution_plan.h"
+
+// Forward declarations.
+void SemiApplyFree(OpBase *opBase);
+OpResult SemiApplyInit(OpBase *opBase);
+Record SemiApplyConsume(OpBase *opBase);
+OpResult SemiApplyReset(OpBase *opBase);
+
+static Record _pullFromBranchStream(OpSemiApply *op) {
+	// Propegate record to the top of the match stream.
+	if(op->op_arg) Argument_AddRecord(op->op_arg, Record_Clone(op->r));
+	return OpBase_Consume(op->match_branch);
+}
+
+static Record OpSemiApply_SemiApplyConsume(OpBase *opBase) {
+	OpSemiApply *op = (OpSemiApply *)opBase;
+	while(true) {
+		// Try to get a record from bound stream.
+		op->r = OpBase_Consume(op->bound_branch);
+		if(!op->r) return NULL; // Depleted.
+
+		// Try to get a record from match stream.
+		Record righthand_record = _pullFromBranchStream(op);
+		if(righthand_record) {
+			// Don't care for matched record.
+			Record_Free(righthand_record);
+			Record r = op->r;
+			op->r = NULL;   // Null to avoid double free.
+			return r;
+		}
+		// Did not managed to get a record from right-hand side, loop back and restart.
+		Record_Free(op->r);
+		op->r = NULL;
+	}
+}
+
+static Record OpSemiApply_AntiSemiApplyConsume(OpBase *opBase) {
```

Comment:
I think removing the `OpSemiApply_` prefix would improve this function name.

Reply:
done